src/NSL/FermionMatrix/Impl/hubbardDiag.hpp

```diff
@@ -43,13 +44,64 @@ class HubbardDiag : public FermionMatrix<Type,LatticeType> {
     *  \param beta  a floating point number where delta=beta/N_t.
     **/
     
-    HubbardDiag(LatticeType & lat,  const NSL::Tensor<Type> &phi, const Type & beta = 1.0 ):
+    HubbardDiag(LatticeType & lat, const NSL::size_t Nt, const Type & beta = 1.0, const Type & mu = 0.0 ):
         FermionMatrix<Type,LatticeType>(lat),
-        phi_(phi),
-        delta_(beta/phi.shape(0)),
-        phiExp_(NSL::LinAlg::exp(phi*NSL::complex<typename NSL::RT_extractor<Type>::value_type>(0,1)))
+        species_(NSL::Hubbard::Species::Particle),
+        delta_( beta/Nt ),
+        mu_( (beta/Nt)*mu ),
+        sgn_( +1 ),
+        phi_( lat.device(), Nt, lat.sites() ),
+        phiExp_( lat.device(), Nt, lat.sites() ),
+        phiExpInv_( lat.device(), Nt, lat.sites() ),
+        Fk_( lat.device(), Nt, lat.sites(), lat.sites() ),
+        FkFkFk_(lat.device(), Nt, lat.sites(), lat.sites()),
+        invAp1F_(lat.device(), lat.sites(), lat.sites()),
+        pi_dot_(lat.device(), Nt, lat.sites())
     {}
 
+    HubbardDiag(NSL::Hubbard::Species species, LatticeType & lat, const NSL::size_t Nt, const Type & beta = 1.0, const Type & mu = 0.0 ):
+        FermionMatrix<Type,LatticeType>(lat),
+        species_(species),
+        delta_( beta/Nt ),
```

Comment:
Does this + surrounding lines need to be on `lat.device()` too?  How can we do that?

Reply:
I don't think so. If lattice is on the device and all incoming Tensors (`phi`) in `populate` are on the device we will just run on GPU.

Reply:
But delta_ doesn't need to be on a particular device?

Reply:
It's just a number, those are copied to device on demand.

Reply:
OK.